File: 프로그래머스/4/43236. 징검다리/징검다리.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

int solution(int distance, vector<int> rocks, int n) {
    rocks.push_back(distance);
    sort(rocks.begin(), rocks.end());
   
    int l = 0, r = distance + 1, m, prev, cnt, answer = 0;
    while(l < r){
        m = (l + r) / 2;
        cnt = 0, prev = 0;
        for(int i = 0; i< rocks.size(); i++){
            if(rocks[i] - prev < m) cnt++;
            else prev = rocks[i];
        }
        if(cnt > n) r = m;
        else {
            answer = max(answer, m);
            l = m + 1;
        }
    }
    return answer;
}

File: 프로그래머스/4/43236. 징검다리/징검다리.java
import java.util.*;

class Solution {
    public int solution(int distance, int[] rocks, int n) {   
        if(rocks.length == n) return distance;
        
        Arrays.sort(rocks);
       
        int lo = 0, hi = distance + 1, m = 0, cnt = 0, prev = 0;
      
        while(lo + 1 < hi) {
            m = (lo + hi) / 2;
            cnt = 0; prev = 0;
            
            for(int r : rocks) {
                if(r - prev < m) cnt++;
                else prev = r;
            }
            
            if(distance - prev < m) cnt++;
            if(cnt > n) hi = m;
            else lo = m;
        }
        
        return lo;
    }
}